http/wsmessage.go

```diff
@@ -0,0 +1,546 @@
+/*
+ * Copyright (C) 2017 Red Hat, Inc.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package http
+
+import (
+	"bytes"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/abbot/go-http-auth"
+	"github.com/golang/protobuf/proto"
+	"github.com/gorilla/websocket"
+	"github.com/nu7hatch/gouuid"
+
+	"github.com/skydive-project/skydive/common"
+	"github.com/skydive-project/skydive/logging"
+)
+
+const (
+	// WildcardNamespace is the namespace used as wildcard. It is used by listeners to filter callbacks.
+	WildcardNamespace = "*"
+	ProtobufProtocol  = "protobuf"
+	JsonProtocol      = "json"
+	DefaultProtocol   = ProtobufProtocol
+)
+
+// DefaultRequestTimeout default timeout used for Request/Reply JSON message.
+var DefaultRequestTimeout = 10 * time.Second
+
+type WSStructMessageJSON struct {
+	Namespace string
+	Type      string
+	UUID      string
+	Status    int64
+	Obj       *json.RawMessage
+}
+
+type WSStructMessage struct {
+	Protocol  string
+	Namespace string
+	Type      string
+	UUID      string
+	Status    int64
+	value     interface{}
```

Comment:
It may make the code cleaner to make value a reference to an interface like:
```
Debug() string
Marshal() []byte
Bytes()
DecodeObj(obj interface{}) error
UnmarshalObj(obj interface{}) error
```

Reply:
good idea

Reply:
Ok. In the simplest case, you could do
```
type JSONValue interface{}
type ProtobufValue interface{}
```
and define the methods.